File: src/go/queue.go
package mrpc

import (
	"sync"
)

type call struct {
	Finish bool
	Result string
	err    error
	done   chan struct{}
}

func newCall() *call {
	return &call{
		done: make(chan struct{}),
	}
}

func (c *call) Set(finish bool, result string, err error) {
	c.Finish = finish
	c.Result = result
	c.err = err
	close(c.done)
}

func (c *call) Wait() {
	<-c.done
}

type clientQueue struct {
	mu    sync.Mutex
	queue map[string]*call
}

func newClientQueue() *clientQueue {
	return &clientQueue{
		queue: make(map[string]*call),
	}
}

func (q *clientQueue) WaitResult(key string) {
	q.mu.Lock()
	defer q.mu.Unlock()
	q.queue[key] = newCall()
}

func (q *clientQueue) SetResult(key string, result string, err error) {
	q.mu.Lock()
	defer q.mu.Unlock()
	call, ok := q.queue[key]
	if !ok {
		panic("key not found in queue")
	}
	call.Set(true, result, err)
}

func (q *clientQueue) GetResult(key string) (string, error) {
	q.mu.Lock()
	call, ok := q.queue[key]
	if !ok {
		panic("key not found in queue")
	}
	q.mu.Unlock()

	call.Wait()
	result := call.Result
	err := call.err

	q.mu.Lock()
	delete(q.queue, key)
	q.mu.Unlock()

	return result, err
}

File: src/python/mrpc/queue.py
import threading


class Call:
    def __init__(self):
        self.finish = False
        self.result = ""
        self.err = None
        self.cond = threading.Condition()  # 条件变量用于等待结果

    def set(self, finish, result, err):
        with self.cond:
            self.finish = finish
            self.result = result
            self.err = err
            self.cond.notify_all()  # 唤醒所有等待者

    def wait(self):
        with self.cond:
            if self.finish:
                return
            while not self.finish:
                self.cond.wait()  # 阻塞等待结果


class ClientQueue:
    def __init__(self):
        self.lock = threading.Lock()
        self.queue = {}

    def wait_result(self, key):
        with self.lock:
            self.queue[key] = Call()

    def set_result(self, key, result, err):
        with self.lock:
            call = self.queue[key]
            call.set(True, result, err)

    def get_result(self, key):
        with self.lock:
            call = self.queue[key]
        # interesting
        call.wait()
        result = call.result
        err = call.err
        del self.queue[key]
        return result, err
